### PR TITLE
Add EnsureOpsAssign traits

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

This PR continues the work on https://github.com/centrifuge/centrifuge-chain/pull/1008, extending the ops with *assign* traits.

Before:
```rust
let mut my_long_variable_name: u32 = 0;
my_long_variable_name = my_long_variable_name.checked_sub(1).ok_or(ArithmeticError::Underflow)?;
```
After:
```rust
let mut my_long_variable_name: u32 = 0;
my_long_variable_name.ensure_sub_assign(1)?;
```

## Changes and Descriptions

Addition of new ops traits into `libs/traits`:
- `EnsureAddAssign`
- `EnsureSubAssign`
- `EnsureMulAssign`
- `EnsureDivAssign`
- Correct multiplication underflow behavior
- Correct addition underflow behavior
- Correct subtraction overflow behavior

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

```sh
cargo test -p cfg-traits // Which will run the associated doc examples
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I rebased on the latest `main` branch

### CODE REVIEW

**Comment by NunoAlexandre** [comment]
Will have a look today 🚀 👀

File: libs/traits/src/lib.rs [hunk 1]
```diff
@@ -776,6 +776,8 @@ pub mod ops {
 		/// assert_eq!(extrinsic(), Err(DispatchError::Arithmetic(ArithmeticError::Overflow)));
 		/// ```
 		fn ensure_mul(&self, v: &Self) -> Result<Self, ArithmeticError> {
+			// TODO: Improve the error checking the case where the multiplication shoul be
```
Comment (NunoAlexandre) [inline libs/traits/src/lib.rs:779]:
```suggestion
			// TODO: Improve the error checking the case where the multiplication should be
```

Comment (NunoAlexandre) [reply libs/traits/src/lib.rs:779]: Alternatively, you can add an error variant `ArithmeticError::OutOfBounds` which would be an union of Overflow and Underflow, which I guess would be ok in terms of error handling?

File: libs/traits/src/lib.rs [hunk 2]
```diff
@@ -806,4 +808,105 @@ pub mod ops {
...
+		fn ensure_div_assign(&mut self, v: &Self) -> Result<(), ArithmeticError> {
+			*self = self.ensure_div(v)?;
+			Ok(())
+		}
+	}
+
+	impl<T: EnsureAdd> EnsureAddAssign for T {}
+	impl<T: EnsureSub> EnsureSubAssign for T {}
+	impl<T: EnsureMul> EnsureMulAssign for T {}
+	impl<T: EnsureDiv> EnsureDivAssign for T {}
```
Comment (NunoAlexandre) [inline libs/traits/src/lib.rs:911]: Sweet 👌

**Review by NunoAlexandre — APPROVED** [review]
Looks good to me 🎸

File: libs/traits/src/lib.rs (hunk 1 above)
Comment (lemunozm) [reply libs/traits/src/lib.rs:779]: Thanks for the review Nuno 😃 

The problem is that `ArithmeticError` is something defined by Substrate 😢 . 
I have an idea in mind. I'm going to add a new commit to it soon.

Comment (lemunozm) [reply libs/traits/src/lib.rs]: Updated. I think I have covered all the cases.

File: libs/traits/src/lib.rs [hunk 3]
```diff
@@ -804,6 +857,120 @@ pub mod ops {
...
+		///
+		/// fn extrinsic() -> DispatchResult {
+		///     let mut max = u32::MAX;
+		///     max.ensure_add_assign(&1)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(extrinsic(), Err(DispatchError::Arithmetic(ArithmeticError::Overflow)));
+		/// ```
+		fn ensure_add_assign(&mut self, v: &Self) -> Result<(), ArithmeticError> {
```
Comment (NunoAlexandre) [inline libs/traits/src/lib.rs:881]: Given that this is generic and could be used with a signed number type, couldn't this operation also lead to an Underflow error case? For example if the left hand side is the MIN of said type and right hand side is a some non-zero negative number?

Comment (NunoAlexandre) [reply libs/traits/src/lib.rs:881]: Same question applies to the `ensure_sub_assign`

**Review by NunoAlexandre — COMMENTED** [review]
One question 👀

File: libs/traits/src/lib.rs (hunk 3 above)
Comment (lemunozm) [reply libs/traits/src/lib.rs:881]: Totally right. I'm going to update it.

**Review by NunoAlexandre — APPROVED** [review]
Looks good to me :)

**Pull request merged**